thruster/examples/custom_middleware_with_auth.rs

```diff
@@ -0,0 +1,101 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use log::info;
+use thruster::{
+    context::typed_hyper_context::TypedHyperContext,
+    errors::{ErrorSet, ThrusterError},
+    hyper_server::HyperServer,
+    m,
+    middleware::cookies::HasCookies,
+    middleware_fn, App, HyperRequest, MiddlewareNext, MiddlewareResult, ThrusterServer,
+};
+use tokio::sync::Mutex;
+
+type Ctx = TypedHyperContext<RequestConfig>;
+
+struct User {
+    first: String,
+    last: String,
+}
+
+#[derive(Default)]
+struct ServerConfig {
+    pool: Arc<Mutex<HashMap<String, User>>>,
+}
+
+#[derive(Default)]
+struct RequestConfig {
+    pool: Arc<Mutex<HashMap<String, User>>>,
+    user: Option<User>,
+}
+
+fn generate_context(request: HyperRequest, state: &ServerConfig, _path: &str) -> Ctx {
+    Ctx::new(
+        request,
+        RequestConfig {
+            pool: state.pool.clone(),
+            user: None,
+        },
+    )
+}
+
+#[middleware_fn]
+async fn hello(mut context: Ctx, _next: MiddlewareNext<Ctx>) -> MiddlewareResult<Ctx> {
+    let user = context.extra.user.as_ref().unwrap();
+
+    context.body(&format!("Hello, {} {}", user.first, user.last));
+
+    Ok(context)
+}
+
+#[middleware_fn]
+async fn authenticate(mut context: Ctx, next: MiddlewareNext<Ctx>) -> MiddlewareResult<Ctx> {
+    let db = context.extra.pool.clone();
+
+    let session_header = context
+        .get_header("Session-Token")
+        .pop()
+        .unwrap_or_default();
+
+    let user = db.lock().await.remove(&session_header);
+
+    match user {
+        Some(user) => {
+            context.extra.user = Some(user);
+            next(context).await
+        }
+        None => Err(ThrusterError::unauthorized_error(context)),
+    }
+}
+
+fn main() {
+    env_logger::init();
+    info!("Starting server...");
+
+    let app = App::<HyperRequest, Ctx, ServerConfig>::create(
+        generate_context,
+        ServerConfig {
```

Comment:
Maybe add a comment explaining that you can add some values to state here, also use from ENV or even pass a default DB Pool

Reply:
Since this is an example, I think extending the functionality is a bad idea. Rather I'm actually in the process of writing a tutorial to get new users started that contains things like shared state, db access, etc. that's a much better place for a complex example.

Reply:
That tutorial would be super useful! Thanks in advance!